### PR TITLE
Add partition dpcpp kernels

### PR DESCRIPTION
**Pull request by MarcelKoch**
This PR adds the missing partition dpcpp kernels. I think the previous issue https://github.com/oneapi-src/oneDPL/pull/388 is resolved.

Needed for #985.

### CODE REVIEW

File: dpcpp/distributed/partition_kernels.dp.cpp [hunk 1]
```diff
@@ -30,25 +30,135 @@ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
...
+            if (i == 0) {
+                // set sentinel value at the end
+                part_ids[num_ranges] = num_parts;
+            }
+            range_sizes[i] = range_offsets[i + 1] - range_offsets[i];
+            part_ids[i] = range_parts[i];
+            permutation[i] = static_cast<GlobalIndexType>(i);
+        },
+        num_ranges, num_ranges, num_parts, range_offsets, range_parts,
+        range_sizes.get_data(), part_ids.get_data(), permutation.get_data());
```
Comment (upsj) [inline dpcpp/distributed/partition_kernels.dp.cpp:76]:
```suggestion
        range_sizes, part_ids, permutation);
```

File: dpcpp/distributed/partition_kernels.dp.cpp [hunk 2]
```diff
@@ -30,25 +30,135 @@ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
...
+                part_sizes[cur_part] = grouped_starting_indices[i];
+            }
+            // write result shifted by one entry to get exclusive prefix sum
+            starting_indices[orig_idxs[i]] =
+                prev_part == cur_part ? grouped_starting_indices[i - 1]
+                                      : LocalIndexType{};
+        },
+        num_ranges, range_sizes.get_const_data(), part_ids.get_const_data(),
+        permutation.get_const_data(), starting_indices, part_sizes);
+}
```
Comment (upsj) [inline dpcpp/distributed/partition_kernels.dp.cpp:108]:
```suggestion
        num_ranges, range_sizes, part_ids,
        permutation, starting_indices, part_sizes);
}
```

File: dpcpp/distributed/partition_kernels.dp.cpp [hunk 3]
```diff
@@ -30,25 +30,135 @@ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
...
                             const comm_index_type* range_parts,
                             size_type num_ranges, comm_index_type num_parts,
                             comm_index_type& num_empty_parts,
                             LocalIndexType* starting_indices,
-                            LocalIndexType* part_sizes) GKO_NOT_IMPLEMENTED;
+                            LocalIndexType* part_sizes)
+{
+    if (num_ranges > 0) {
+        auto policy =
+            oneapi::dpl::execution::make_device_policy(*exec->get_queue());
```
Comment (upsj) [inline dpcpp/distributed/partition_kernels.dp.cpp:125]: Maybe we can add something like `thrust.cuh` taking care of this policy setup?

**Review by upsj — APPROVED** [review]
LGTM! Only small suggestions. I guess we can now reactivate the DPC++ `mpi/matrix` tests?

File: dpcpp/base/kernel_launch_reduction.dp.hpp [hunk 4]
```diff
@@ -194,8 +194,8 @@ void run_kernel_reduction_impl(std::shared_ptr<const DpcppExecutor> exec,
     } else {
         queue->submit([&](sycl::handler& cgh) {
             generic_kernel_reduction_1d<DeviceConfig>(
-                cgh, static_cast<int64>(size), num_workgroups, fn, op, finalize,
-                identity, result, args...);
+                cgh, static_cast<int64>(size), 1, fn, op, finalize, identity,
+                result, args...);
```
[the remark this answers is not in the document]
Comment (MarcelKoch) [reply dpcpp/base/kernel_launch_reduction.dp.hpp:198]: I've put that into https://github.com/ginkgo-project/ginkgo/pull/1362. Somehow I had these changes in a branch for a while, but never put up a PR...

**Review by yhmtsai — APPROVED** [review]
I think it needs to wait for #1362

**Comment by MarcelKoch** [comment]
format-rebase!

**Comment by MarcelKoch** [comment]
rebase!

**Comment by MarcelKoch** [comment]
rebase!

**Comment by MarcelKoch** [comment]
rebase!

**Pull request merged**